## Data Cleaning: IMDB and Movie DB

In [1]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns

In [2]:
# read in datasets for cleaning 
bom_movie_gross_df = pd.read_csv('../../data/01_raw/bom.movie_gross.csv.gz')
tn_movie_budgets_df = pd.read_csv('../../data/01_raw/tn.movie_budgets.csv.gz')
imdb_name_basics_df = pd.read_csv('../../data/01_raw/imdb.name.basics.csv.gz')
imdb_title_akas_df = pd.read_csv('../../data/01_raw/imdb.title.akas.csv.gz')
imdb_title_basics_df = pd.read_csv('../../data/01_raw/imdb.title.basics.csv.gz')
imdb_title_crew_df = pd.read_csv('../../data/01_raw/imdb.title.crew.csv.gz')
imdb_title_principals_df = pd.read_csv('../../data/01_raw/imdb.title.principals.csv.gz')
imdb_title_ratings_df = pd.read_csv('../../data/01_raw/imdb.title.ratings.csv.gz')
tmdb_movies_df = pd.read_csv('../../data/01_raw/tmdb.movies.csv.gz')

### Clean with tn_movie_budgets_df

In [3]:
tn_movie_budgets_df.head()

id  release_date                                        movie  \
0   1  Dec 18, 2009                                       Avatar   
1   2  May 20, 2011  Pirates of the Caribbean: On Stranger Tides   
2   3   Jun 7, 2019                                 Dark Phoenix   
3   4   May 1, 2015                      Avengers: Age of Ultron   
4   5  Dec 15, 2017            Star Wars Ep. VIII: The Last Jedi   

  production_budget domestic_gross worldwide_gross  
0      $425,000,000   $760,507,625  $2,776,345,279  
1      $410,600,000   $241,063,875  $1,045,663,875  
2      $350,000,000    $42,762,350    $149,762,350  
3      $330,600,000   $459,005,868  $1,403,013,963  
4      $317,000,000   $620,181,382  $1,316,721,747

Check for duplicated rows

In [5]:
tn_movie_budgets_df.duplicated('movie').sum()

84

In [6]:
tn_movie_budgets_df['repeated_name'] = tn_movie_budgets_df.movie.duplicated(keep=False)

In [7]:
tn_movie_budgets_df[tn_movie_budgets_df['repeated_name'] == True].sort_values('movie')

id  release_date                         movie production_budget  \
4270   71  Dec 23, 1954  20,000 Leagues Under the Sea        $5,000,000   
5614   15  Dec 24, 1916  20,000 Leagues Under the Sea          $200,000   
1648   49  Apr 30, 2010     A Nightmare on Elm Street       $35,000,000   
5016   17   Nov 9, 1984     A Nightmare on Elm Street        $1,800,000   
2032   33  Nov 11, 1992                       Aladdin       $28,000,000   
80     81  May 24, 2019                       Aladdin      $182,000,000   
50     51   Mar 5, 2010           Alice in Wonderland      $200,000,000   
4759   60  Jul 28, 1951           Alice in Wonderland        $3,000,000   
4120   21  Oct 17, 1956   Around the World in 80 Days        $6,000,000   
340    41  Jun 16, 2004   Around the World in 80 Days      $110,000,000   
134    35  Mar 17, 2017          Beauty and the Beast      $160,000,000   
2485   86  Nov 13, 1991          Beauty and the Beast       $20,000,000   
3007    8  Nov 18, 1959                       Ben-Hur       $15,000,000   
422    23  Aug 19, 2016                       Ben-Hur       $95,000,000   
3307    8  Aug 14, 2015                      Brothers       $13,000,000   
2140   41   Dec 4, 2009                      Brothers       $26,000,000   
4058   59  Mar 26, 2010                          Ca$h        $7,000,000   
3231   32  Dec 31, 2008                          Ca$h       $13,700,000   
1925   26  Oct 18, 2013                        Carrie       $30,000,000   
5015   16  Nov 16, 1976                        Carrie        $1,800,000   
359    60  Nov 17, 2006                 Casino Royale      $102,000,000   
3381   82  Apr 28, 1967                 Casino Royale       $12,000,000   
3325   26   Apr 2, 1982                    Cat People       $12,500,000   
5668   69  Nov 16, 1942                    Cat People          $134,000   
4775   76  Feb 15, 1950                    Cinderella        $2,900,000   
412    13  Mar 13, 2015                    Cinderella       $95,000,000   
3050   51  Jun 12, 1981           Clash of the Titans       $15,000,000   
271    72   Apr 1, 2010           Clash of the Titans      $125,000,000   
475    76  Aug 19, 2011           Conan the Barbarian       $90,000,000   
2543   44  May 14, 1982           Conan the Barbarian       $20,000,000   
...   ...           ...                           ...               ...   
1376   77  Jun 11, 2010                The Karate Kid       $40,000,000   
3862   63  Jun 22, 1984                The Karate Kid        $8,000,000   
3043   44  Mar 13, 2009    The Last House on the Left       $15,000,000   
5699  100  Aug 30, 1972    The Last House on the Left           $87,000   
5603    4  Aug 18, 2015               The Love Letter          $250,000   
3109   10  May 21, 1999               The Love Letter       $15,000,000   
64     65   Jun 9, 2017                     The Mummy      $195,000,000   
543    44   May 7, 1999                     The Mummy       $80,000,000   
2189   90   Jun 6, 2006                      The Omen       $25,000,000   
4784   85  Jun 25, 1976                      The Omen        $2,800,000   
5718   19  Feb 22, 2008                    The Signal           $50,000   
4525   26  Jun 13, 2014                    The Signal        $4,000,000   
5099  100  Oct 25, 2013                    The Square        $1,500,000   
5009   10   Apr 9, 2010                    The Square        $1,900,000   
3233   34   Oct 5, 1956          The Ten Commandments       $13,500,000   
3694   95  Oct 19, 2007          The Ten Commandments       $10,000,000   
3768   69  Oct 17, 2003   The Texas Chainsaw Massacre        $9,000,000   
5665   66  Oct 18, 1974   The Texas Chainsaw Massacre          $140,000   
779    80   Jun 1, 1990                  Total Recall       $65,000,000   
278    79   Aug 3, 2012                  Total Recall      $125,000,000   
5330   31  Dec 31, 2012                        Trance          $950,000   
2970   71   Apr 5, 2013         

There appear to be no actually duplicated data, all of the duplicated rows appear to be remakes of an original. Lets change the name of that column from ```repeated_name``` to ```remade```.

In [8]:
tn_movie_budgets_df['repeated_name'] = tn_movie_budgets_df.movie.duplicated()

In [9]:
tn_movie_budgets_df.rename(columns={'repeated_name':'remade'}, inplace=True)

***
### Are their weird values?
***

While checking for duplicates, we noticed that there were 0s in the ```domestic_gross``` and ```worldwide_gross``` columns.
Lets do the following:
- [ ] first, convert all money columns to ints from objects, 
- [ ] second, make a new column called ```international_gross``` which is ```worldwide_gross```$-$```domestic_gross```, and
- [ ] finally, decide whether or not to drop rows with too many \$0s.
***

In [10]:
tn_movie_budgets_df.dtypes

id                    int64
release_date         object
movie                object
production_budget    object
domestic_gross       object
worldwide_gross      object
remade                 bool
dtype: object

In [11]:
tn_movie_budgets_df = tn_movie_budgets_df.astype({'production_budget':'str', 'domestic_gross':'str', 'worldwide_gross':'str'})

The following 5 cells should be rolled into one function called convert_money_obj_to_int

In [12]:
def get_rid_of_dollar_sign(amount):
    if amount.startswith('$'):
        amount = amount[1:]
    else:
        print('Crap, one of my values didn\'t start with a dollar sign')
    return amount

In [13]:
for title in ['production_budget', 'domestic_gross', 'worldwide_gross']:
    tn_movie_budgets_df[title] = tn_movie_budgets_df[title].map(get_rid_of_dollar_sign)

In [14]:
for title in ['production_budget', 'domestic_gross', 'worldwide_gross']:
    tn_movie_budgets_df[title] = tn_movie_budgets_df[title].map(lambda x: x.replace(',','_'))

In [15]:
tn_movie_budgets_df.head()

id  release_date                                        movie  \
0   1  Dec 18, 2009                                       Avatar   
1   2  May 20, 2011  Pirates of the Caribbean: On Stranger Tides   
2   3   Jun 7, 2019                                 Dark Phoenix   
3   4   May 1, 2015                      Avengers: Age of Ultron   
4   5  Dec 15, 2017            Star Wars Ep. VIII: The Last Jedi   

  production_budget domestic_gross worldwide_gross  remade  
0       425_000_000    760_507_625   2_776_345_279   False  
1       410_600_000    241_063_875   1_045_663_875   False  
2       350_000_000     42_762_350     149_762_350   False  
3       330_600_000    459_005_868   1_403_013_963   False  
4       317_000_000    620_181_382   1_316_721_747   False

In [16]:
tn_movie_budgets_df = tn_movie_budgets_df.astype({'production_budget':'int64', 'domestic_gross':'int64', 'worldwide_gross':'int64'})

In [17]:
tn_movie_budgets_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5782 entries, 0 to 5781
Data columns (total 7 columns):
id                   5782 non-null int64
release_date         5782 non-null object
movie                5782 non-null object
production_budget    5782 non-null int64
domestic_gross       5782 non-null int64
worldwide_gross      5782 non-null int64
remade               5782 non-null bool
dtypes: bool(1), int64(4), object(2)
memory usage: 276.8+ KB


***
- [x] first, convert all money columns to ints from objects, 
- [ ] second, make a new column called ```international_gross``` which is ```worldwide_gross```$-$```domestic_gross```, and
- [ ] finally, decide whether or not to drop rows with too many \$0s.
***

While we're at it, we might as well make all the columns the appropriate data types...

In [19]:
tn_movie_budgets_df = tn_movie_budgets_df.astype({'movie':'str'})

In [20]:
tn_movie_budgets_df.release_date = pd.to_datetime(tn_movie_budgets_df.release_date)

In [21]:
tn_movie_budgets_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5782 entries, 0 to 5781
Data columns (total 7 columns):
id                   5782 non-null int64
release_date         5782 non-null datetime64[ns]
movie                5782 non-null object
production_budget    5782 non-null int64
domestic_gross       5782 non-null int64
worldwide_gross      5782 non-null int64
remade               5782 non-null bool
dtypes: bool(1), datetime64[ns](1), int64(4), object(1)
memory usage: 276.8+ KB


...Okay, back to the business at hand.

In [22]:
tn_movie_budgets_df['international_gross'] = tn_movie_budgets_df['worldwide_gross'] - tn_movie_budgets_df['domestic_gross']

- [x] first, onvert all money columns to ints from objects, 
- [x] second, make a new column called ```international_gross``` which is ```worldwide_gross```$-$```domestic_gross```, and
- [ ] finally, decide whether or not to drop rows with too many \$0s.

In [23]:
tn_movie_budgets_df.loc[tn_movie_budgets_df.international_gross == 0].shape

(1619, 8)

***
Thats a lot of movies with no international gross!

Now I'm concerned about movies with no worldwide gross or no domestic gross. 
Lets see how many of those there are.
***

In [24]:
tn_movie_budgets_df.loc[(tn_movie_budgets_df.domestic_gross) == 0].shape

(548, 8)

In [25]:
tn_movie_budgets_df.loc[(tn_movie_budgets_df.domestic_gross) == 0].head()

id release_date                                           movie  \
194  95   2020-12-31                                        Moonfall   
479  80   2017-12-13                                          Bright   
480  81   2019-12-31                                Army of the Dead   
535  36   2020-02-21                                Call of the Wild   
617  18   2012-12-31  Astérix et Obélix: Au service de Sa Majesté   

     production_budget  domestic_gross  worldwide_gross  remade  \
194          150000000               0                0   False   
479           90000000               0                0   False   
480           90000000               0                0   False   
535           82000000               0                0   False   
617           77600000               0         60680125   False   

     international_gross  
194                    0  
479                    0  
480                    0  
535                    0  
617             60680125

In [26]:
tn_movie_budgets_df.loc[(tn_movie_budgets_df.worldwide_gross | tn_movie_budgets_df.domestic_gross) == 0].shape

(367, 8)

In [27]:
tn_movie_budgets_df.loc[(tn_movie_budgets_df.worldwide_gross | tn_movie_budgets_df.domestic_gross) == 0].head()

id release_date             movie  production_budget  domestic_gross  \
194  95   2020-12-31          Moonfall          150000000               0   
479  80   2017-12-13            Bright           90000000               0   
480  81   2019-12-31  Army of the Dead           90000000               0   
535  36   2020-02-21  Call of the Wild           82000000               0   
670  71   2019-08-30         PLAYMOBIL           75000000               0   

     worldwide_gross  remade  international_gross  
194                0   False                    0  
479                0   False                    0  
480                0   False                    0  
535                0   False                    0  
670                0   False                    0

In [28]:
tn_movie_budgets_df.loc[(tn_movie_budgets_df.domestic_gross) == 0].loc[(tn_movie_budgets_df.worldwide_gross) > 0].shape

(181, 8)

In [29]:
tn_movie_budgets_df.loc[(tn_movie_budgets_df.domestic_gross) == 0].loc[(tn_movie_budgets_df.worldwide_gross) > 0].head()

id release_date                                           movie  \
617   18   2012-12-31  Astérix et Obélix: Au service de Sa Majesté   
619   20   2019-01-22                                       Renegades   
820   21   2018-10-26                                      Air Strike   
1325  26   2012-12-31                                      Foodfight!   
1367  68   2006-12-31                Les Bronzés 3: amis pour la vie   

      production_budget  domestic_gross  worldwide_gross  remade  \
617            77600000               0         60680125   False   
619            77500000               0          1521672   False   
820            65000000               0           516279   False   
1325           45000000               0            73706   False   
1367           42000000               0         83833602   False   

      international_gross  
617              60680125  
619               1521672  
820                516279  
1325                73706  
1367             83833602

In [30]:
tn_movie_budgets_df.loc[(tn_movie_budgets_df.worldwide_gross) == 0].shape

(367, 8)

In [31]:
tn_movie_budgets_df.loc[(tn_movie_budgets_df.worldwide_gross) == 0].head()

id release_date             movie  production_budget  domestic_gross  \
194  95   2020-12-31          Moonfall          150000000               0   
479  80   2017-12-13            Bright           90000000               0   
480  81   2019-12-31  Army of the Dead           90000000               0   
535  36   2020-02-21  Call of the Wild           82000000               0   
670  71   2019-08-30         PLAYMOBIL           75000000               0   

     worldwide_gross  remade  international_gross  
194                0   False                    0  
479                0   False                    0  
480                0   False                    0  
535                0   False                    0  
670                0   False                    0

There are 548 movies with 0 dollars in ```domestic_gross```.
Of those movies, 181 have ```international_gross``` (maybe were only released internationally), and the other 367 have no ```domestic_gross``` or ```international_gross```. 
We think these 367 movies were released online only, which means they are of particular interest to our analysis!
***
So we are deciding to keep all of our data:
- [x] first, onvert all money columns to ints from objects, 
- [x] second, make a new column called ```international_gross``` which is ```worldwide_gross```$-$```domestic_gross```, and
- [x] finally, decide whether or not to drop rows with too many \$0s.


We are also deciding to compare the set of data we think are online releases to the data we think are not. 
Let's create a new column to mark their differences, then continue cleaning the data by isolating the released between 2010 and 2018.

In [33]:
tn_movie_budgets_df['online_release'] = tn_movie_budgets_df.worldwide_gross.map(lambda x: x==0)

### Get rid of movies not released between 2010 and 2018=

In [34]:
tn_movie_budgets_df['release_year'] = tn_movie_budgets_df.release_date.map(lambda x: x.year)

In [35]:
recent_tn_movie_budgets_df = tn_movie_budgets_df.loc[(2010<=tn_movie_budgets_df['release_year']) & (tn_movie_budgets_df['release_year']<=2018)]

In [36]:
recent_tn_movie_budgets_df.online_release.value_counts()

False    1873
True      251
Name: online_release, dtype: int64

There are still 251 online releases and 1873 box office releases; enough to perform some analysis.

In [37]:
recent_tn_movie_budgets_df.to_csv('../../data/02_intermediate/recent_tn_movie_budgets.csv', index=False)